### PR TITLE
Add Experimental Library for Calling Cute/Cutlass DSL Kernels in Jax

### PR DESCRIPTION
**Pull request by mgoldfarb-nvidia**
This PR contains the first implementation of our integration of Cute DSL into Jax.

### CODE REVIEW

**Review by gpupuck — DISMISSED** [review]
Looks good to me!

File: .github/container/jax_cutlass/README.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    constval=1.0,
+    ...
+)
+```
+
+## Layouts and Modes
+
+`cutlass_call` accepts two optional parameters to aid in converting `jax.Array` into a `cute.Tensor`:
+
+_Layout_: Describes the physical order of axis strides for the source `jax.Array`. If omitted the array is assumed to have row-major layout.
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/README.md]: Are these always tuples? It also seems that you specify the layout and/or mode for inputs and outputs, could you spell that out here?

File: .github/container/jax_cutlass/README.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+A common example of when to use modes is to implement logical layouts for gemm operations. For example we want the `cute.Tensor`s to follow a consistent layout specification of `[M][K][L]` for A, `[N][K][L]` for B and `[M][N][L]` for C. Strides of each dimension are set to properly offset into physical memory. 
+
+
+### Complex Layouts and Modes
+
+If your kernel requires more complex layouts at the function boundary e.g. tiled, composed or hierarchical it is recommended that the kernel be wrapped in a `cute.jit` function. Once you are outside of Jax you can built arbitrary layouts as needed using CuTe DSL.
+
+## Compilation Cache
+
+Like Jax, we maintain a cache of the compiled functions. When a kernel is compiled we check if it was previously seen for the given shapes, dtypes and constant values. For example if your kernel is called in a series of homogeneous layer (e.g a transformer model) it will only need to compile once and that instance can be reused as needed.
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/README.md]: Is there a user-configurable directory that can be reused across ranks? If not, then I would perhaps not imply that it is "Like JAX".

File: .github/container/jax_cutlass/README.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+## Limitations
+
+There are several limitations to highlight to avoid unexpected errors or behavior. Over time we hope to improve these as CuTe DSL matures.
+
+### Jit Function Argument Types
+
+`cutlass_call` does not allow for arbitrary python types to be passed between Jax and the DSL kernel. If the kernel interface depends on a complex Python type it is recommended that a wrapper function be used to bind together `jax.Array` and other compile time constants that can be passed.
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/README.md]: Please spell out which types are allowed. Only `jax.Array`s? Or also lists and tuples of `jax.Arrays`?

File: .github/container/jax_cutlass/README.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+@cute.jit
+def wrapper(stream: cuda.CUstream, a: cute.Tensor, b: cute.Tensor, *, constval: float):
+    x = CustomType(a, b, constval)
+    launch(stream, x)
+
+out = cutlass_call(wrapper, ..., constval=1.0)(a, b)
+```
+
+A useful trick can be to flatten a `PyTree` structure outside the call then unflatten inside the call by passing the `PyTree` as a constexpr argument.
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/README.md]: Is there a reason why `cutlass_call` does not do that automatically? (Similarly to `buffer_callback`.)

File: .github/container/jax_cutlass/README.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+    @cute.jit
+    def __call__(self, stream, ...):
+        # use self.v0, self.v1
+        ...
+
+def make_function(v0, v1):
+    return MyFunction(v0, v1)
+```
+
+One common exception is if you need to
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/README.md]: Incomplete sentence.

File: .github/container/jax_cutlass/jax_cutlass/primitive.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    input_mode_flat: Any,
+    output_mode_flat: Any,
+    input_output_aliases,
+    convert_tensors,
+    allow_cuda_graph,
+    **kwargs,
+):
+    input_output_aliases = dict(input_output_aliases)
+
+    # TODO: Need to support device-less compilation and deffer module load
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/jax_cutlass/primitive.py]: Nit: s/deffer/defer/

File: .github/container/jax_cutlass/jax_cutlass/primitive.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+def _cutlass_call_vmap_rule(*args, **kwargs):
+    del args, kwargs
+    raise NotImplementedError(
+        "cutlass_call does not support batching with jax.vmap. Please "
+        "use jax.custom_batching.custom_vmap for applying vmap. "
+    )
+
+
+batching.primitive_batchers[cutlass_call_inner_p] = _cutlass_call_vmap_rule
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/jax_cutlass/primitive.py]: Suggestion for future: Perhaps you could wire up vmap_method from JAX FFI/buffer_callback.

File: .github/container/jax_cutlass/tests/test_args.py [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+    for ci, ci_ref in zip(c, c_ref):
+        assert jnp.allclose(ci, ci_ref)
+
+
+def test_list_args_alias():
+
+    @cute.kernel
+    def kernel(
+        a: cute.Tensor,
+        b: list[cute.Tensor],
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/tests/test_args.py]: How about also testing tuple args?

File: .github/container/jax_cutlass/README.md (hunk 1 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/README.md]: Yes these are tuples currently. More complex layouts/orders could require custom types. I think we could also accept `Layout` from jax and normalize/convert it. Tuples are at least safe choice for now I think.

Do you have any thoughts on `Layout` and its relation here?

I will update the wording to reflect these are specified individually for both input and output even when there is aliasing.

File: .github/container/jax_cutlass/README.md (hunk 2 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/README.md]: The cache is in memor- we don't save it to disk although that is a feature I'd like to add support for in a follow up.

I'll remove `Like Jax` to avoid confusion.

File: .github/container/jax_cutlass/README.md (hunk 5 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/README.md]: Ooops will finish this one :)

File: .github/container/jax_cutlass/tests/test_args.py (hunk 8 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/tests/test_args.py]: Good suggestion, I will make one of the `list` a `tuple`.

File: .github/container/jax_cutlass/README.md (hunk 4 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/README.md]: We actually do automatically flatten and unflatten to support list types. But I left more complicated mapping to the user. Perhaps this last sentence will just cause confusion?

There are a few challenges that make it tricky and why I left it as a future feature:
1. There are limitations to what types can pass the jit boundry and a custom protocol is used to augment user defined types. Primitive types like dict are not allowed although it may be possible to wrap/inherit. We might be able to provide utilities to augment dataclasses and provide documentation for adventurous users.

2. Input / Output representation in a Pytree gets tricky since you may need to alias arrays in the tree if the tree combines both inputs and outputs (e.g. an update operation). In this case it might be better to fully alias the entire tree than try to return a new tree with updates that get merged in somehow.

The current simplified approach basically says if you have some complex pytree structure you'd need to do the work to flatten it prior to calling cutlass kernel. But I agree it would be nice to support more natively arbitrary Pytrees but its going to take some more work.

File: .github/container/jax_cutlass/jax_cutlass/primitive.py (hunk 7 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/jax_cutlass/primitive.py]: Agree will look at this for next update.

File: .github/container/jax_cutlass/README.md (hunk 3 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/README.md]: Done. Let me know if you fell this still needs to be made more clear.

File: .github/container/jax_cutlass/README.md (hunk 4 above)
Comment (jaro-sevcik) [reply .github/container/jax_cutlass/README.md]: Makes sense, thank you for the clarification.

File: .github/container/jax_cutlass/README.md (hunk 1 above)
Comment (jaro-sevcik) [reply .github/container/jax_cutlass/README.md]: I see - I did not mean if the individual layouts are tuples, I meant if the structure of layouts is also tuples.

For example, if the call takes a 2-element list of arrays and one other array, would the input_layout be something like `input_layout = ([(0, 1), (2, 1, 0)], (0, 1, 2))` or rather `input_layout = (((0, 1), (2, 1, 0)), (0, 1, 2))` (or does not it matter)?

Also, it would be nice to have some tests for the layouts, if only for documentation purposes.

File: .github/container/jax_cutlass/README.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+            block=[a.shape[-1], 1, 1],
+            stream=stream)
+
+call = cutlass_call(launch, output_shape_dtype=jax.ShapeDtypeStruct((128, 64), jnp.float16))
+out = call(input)
+```
+
+## Host Function Signature
+
+`cutlass_call` requires a specific function signature to bind arrays and constant values provided by Jax. It is recommend that you annotate the signature with the appropriate or expected types.
```
Comment (jaro-sevcik) [inline .github/container/jax_cutlass/README.md]: s/recommend/recommended/

File: .github/container/jax_cutlass/README.md (hunk 3 above)
Comment (jaro-sevcik) [reply .github/container/jax_cutlass/README.md]: I was wondering if you should also say that each argument can also be a list of tensors or a tuple of tensors, but perhaps that is too much information for the docs.

**Review by jaro-sevcik — DISMISSED** [review]

File: .github/container/jax_cutlass/README.md (hunk 1 above)
Comment (mgoldfarb-nvidia) [reply .github/container/jax_cutlass/README.md]: Oh I see, let me double check, I think today it may assume a flattened list.
